### ISSUE TITLE
tf.keras.applications.ResNet50 is not working

### ISSUE THREAD

**Issue by irum** [tensorflow/tensorflow#36915]
<em>I am working on transfer learning and used the ResNet50 model to predict 10 classes of my dataset. The losss went to 0.7 and acc=99% during training phase, but when i evaluate the model on test dataset, it gave me acc=10% and loss=2.275. The most panicking thing was that model was predicting only one class during test phase.  I tried every solution , even for over fitting but in vain. Then, i used ResNet50V2 model instead. This model is predicting every class now and acc=88% during test phase. </em>

**System information** 
- Have I written custom code (as opposed to using a stock
example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g.,
Linux Ubuntu 16.04): Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if
the issue happens on mobile device:  No
- TensorFlow installed from (source or
binary): - TensorFlow version (use command below):  Command version
- Python version: - Bazel
version (if compiling from source):
- GCC/Compiler version (if compiling from
source): 
- CUDA/cuDNN version: - GPU model and memory: 10.1 and cuDNN 7.6

 Please can you check the ResNet50 code as i think there is some problem in it as same code of mine is working with tf.keras.application.ResNet50V2?

Thank you

**Comment by amahendrakar** [comment]
@irum,
I tried transfer learning from [this](https://www.tensorflow.org/tutorials/images/transfer_learning) Tensorflow tutorial using ResNet50 and was able to run the code without any issues. Please find the Gist [here](https://colab.research.google.com/gist/amahendrakar/2c4720f92ec19824d324ef2434f6b250/36915.ipynb). Thanks!

**Comment by irum** [comment]
I have done the coding in the same way by using the above mentioned tutorial but the model was predicting only one class . I used tape.gradient loop to train the model . Then I used model.fit as well to train the model but for my dataset this resent model was not working at all.

**Comment by amahendrakar** [comment]
@irum,
Could you please provide the complete code to reproduce the issue reported here. Thanks!

**Comment by irum** [comment]
`import tensorflow as tf
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras import optimizers,regularizers
import os
from PIL import Image
import numpy as np

Dir = "/media/irum/DATA/Dataset"
Dir_test="/media/irum/DATA/Dataset"

catogories = ['bottles', 'cans', 'chain', 'cloth', 'gloves', 'metal_objects', 'pipe_joints', 'plastic_pipes', 'sponges',
              'wood']

class_names = catogories


train_dir = os.path.join(Dir ,'train')
validation_dir = os.path.join(Dir_test, 'test')

batch_size = 32
epochs =50
IMG_HEIGHT = 224
IMG_WIDTH = 224

total_train= 2000
total_val=1494

train_image_generator = ImageDataGenerator(rescale=1./255) # Generator for our training data
validation_image_generator = ImageDataGenerator(rescale=1./255) # Generator for our validation data

train_data_gen = train_image_generator.flow_from_directory(batch_size=batch_size,
                                                           directory=train_dir,
                                                           shuffle=True,
                                                           target_size=(IMG_HEIGHT, IMG_WIDTH),
                                                           class_mode='categorical')

val_data_gen = validation_image_generator.flow_from_directory(batch_size=batch_size,
                                                              directory=validation_dir,
                                                              target_size=(IMG_HEIGHT, IMG_WIDTH),
                                                              class_mode='categorical')

base_model = tf.keras.applications.resnet50.ResNet50(input_shape=(224, 224, 3), include_top=False, weights='imagenet')

feature_batch = base_model.output


global_average_layer = tf.keras.layers.GlobalAveragePooling2D()
feature_batch_average = global_average_layer(feature_batch)
feature_batch_average=tf.keras.layers.Dropout(0.9)(feature_batch_average)
prediction_layer = tf.keras.layers.Dense(10,activation='softmax', kernel_regularizer=regularizers.l2(0.01))
prediction_batch = prediction_layer(feature_batch_average)

model = tf.keras.Sequential([
    base_model,
    global_average_layer,
    prediction_layer
])

base_learning_rate = 0.0001
model.compile(optimizer=tf.keras.optimizers.RMSprop(lr=base_learning_rate),
              loss=tf.keras.losses.CategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

history = model.fit_generator(
    train_data_gen,
    steps_per_epoch=total_train // batch_size,
    epochs=epochs,
    validation_data=val_data_gen,
    validation_steps=total_val // batch_size
)


img_gen = tf.keras.preprocessing.image.ImageDataGenerator(rescale=1./255)
images, labels = next(img_gen.flow_from_directory(validation_dir))
images=tf.image.resize(images,[224,224])
labels[0]
probability_model = tf.keras.Sequential([model,
                                         tf.keras.layers.Softmax()])
predictions = probability_model.predict(images)
np.argmax(predictions[1])


image_batch = images
label_batch = labels

predicted_batch = model.predict(image_batch)
predicted_batch = tf.squeeze(predicted_batch).numpy()

predicted_ids = np.argmax(predicted_batch, axis=-1)
predicted_ids-np.array(predicted_ids)
print(predicted_ids)
for i in predicted_ids:
    predicted_class_names = class_names[i]
    print("{} ,\t{}".format(predicted_class_names, class_names[np.argmax(labels)])) `


The results after predictions are
[0 1 9 1 0 5 6 6 4 6 5 7 8 8 3 3 7 5 8 0 5 5 1 4 3 7 0 3 8 2 8 4]
bottles ,	bottles
cans ,	bottles
wood ,	bottles
cans ,	bottles
bottles ,	bottles
metal_objects ,	bottles
pipe_joints ,	bottles
pipe_joints ,	bottles
gloves ,	bottles
pipe_joints ,	bottles
metal_objects ,	bottles
plastic_pipes ,	bottles
sponges ,	bottles
sponges ,	bottles
cloth ,	bottles
cloth ,	bottles
plastic_pipes ,	bottles
metal_objects ,	bottles
sponges ,	bottles
bottles ,	bottles
metal_objects ,	bottles
metal_objects ,	bottles
cans ,	bottles
gloves ,	bottles
cloth ,	bottles
plastic_pipes ,	bottles
bottles ,	bottles
cloth ,	bottles
sponges ,	bottles
chain ,	bottles
sponges ,	bottles
gloves ,	bottles

As you can see all predictions are equal to bottles only . 
P.S I had added before dropout layers and regularization layers but the results were same, So no overfitting problem fits here.

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with sample dataset. Please find the Gist [here](https://colab.sandbox.google.com/gist/amahendrakar/b643887f3e32ecb0267dfe4da6e2f4c3/36915.ipynb). Thanks!

**Comment by liuxingbaoyu** [comment]
Thank you! I may have encountered a similar problem and succeeded after I changed basenet.
When I am using mobilenetv2, everything is normal during training, but the embed is all the same during validation, and the predict is also the same (all use the same samples)

**Comment by irum** [comment]
@liuxingbaoyu   
No worries. I am happy that this post was helpful .I already explained to tensorflow team that the base models have some ii sues.

**Comment by liuxingbaoyu** [comment]
This may be related to BatchNormalization.
I use the model without any BatchNormalization and it will be very good.

**Comment by ymodak** [comment]
This question is better asked on [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a bug or feature request. There is also a larger community that reads questions there.
Thanks!

**Comment by liuxingbaoyu** [comment]
@ymodak Maybe related to [this](https://github.com/tensorflow/tensorflow/issues/36700)

**Comment by raghavab1992** [comment]
@irum  try this [solution ](https://github.com/tensorflow/tensorflow/issues/36366#issuecomment-601985968) as BatchNorm might be the issue here

**Comment by irum** [comment]
@raghavab1992 
Thank you Raghav. It is  working in this way.

**Comment by irum** [comment]
it is working but training is very slow in this way. After 500 epochs , i am getting 64% accuracy on training dataset.

**Comment by irum** [comment]
@ymodak , please can you use any other custom dataset for Resnet50. It may work well for cats and dogs dataset.

**Comment by DanielDimanov** [comment]
@raghavab1992 Thank you!!! I've been struggling with this for days and thought my data was corrupted or the head I was putting was existentially wrong. The solution you mentioned worked for me!
